In [117]:
# Import required libraries and dependencies
import pandas as pd

#hvplot.pandas is an extension for pandas that enables interactive plotting using the Holoviews library.
import hvplot.pandas

# KMeans is an implementation of the K-means clustering algorithm,
#  a popular unsupervised machine learning algorithm for clustering 
from sklearn.cluster import KMeans 

#PCA is used for Principal Component Analysis,
#  a dimensionality reduction technique.
from sklearn.decomposition import PCA

#StandardScaler is used to standardize the features by removing the mean and scaling to unit variance.
# It's often essential for clustering algorithms.
from sklearn.preprocessing import StandardScaler

import warnings


warnings.filterwarnings("ignore")



# import pandas as pd
# import hvplot.pandas
# from sklearn.cluster import KMeans
# from sklearn.decomposition import PCA
# from sklearn.preprocessing import StandardScaler

In [118]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [119]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [120]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [121]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
# first lets get the name of the cloumns 
df_market_data.columns

scaled_data = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y']])
scaled_data[:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [122]:
# Create a DataFrame with the scaled data
df_scaled_data = pd.DataFrame(
    scaled_data , columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'] 
)

# Copy the crypto names from the original data
df_scaled_data["coin_id"] = df_market_data.index

# Set the coinid column as index

df_scaled_indexed = df_scaled_data.set_index("coin_id")
# Display sample data

df_scaled_indexed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [123]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))
df_scaled_indexed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [124]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k :
    
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=3)
    
    # 2. Fit the model to the data using 'df_scaled_data'
    k_model.fit(df_scaled_indexed)

    # 3. Append the model.inertia_ to the inertia list

    inertia.append(k_model.inertia_)
inertia

[287.0,
 195.82021818036043,
 123.19048183836956,
 79.02243535120975,
 63.85866780584265,
 53.05778846567061,
 44.550278020774805,
 36.791069782922925,
 32.71517024852882,
 28.374754158229536]

In [125]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k" : k,
    "inertia" : inertia
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_data = pd.DataFrame(elbow_data)
df_elbow_data

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   63.858668
5   6   53.057788
6   7   44.550278
7   8   36.791070
8   9   32.715170
9  10   28.374754

In [126]:
# Plot a line chart with all the inertia values computed with 
df_elbow_data.hvplot.line(
    x = "k",
    y = "inertia",
    title = "Elbow method"
)


# the different values of k to visually identify the optimal value for k.
# df_elbow_data["inertia"]



:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  Four

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [127]:
# Initialize the K-Means model using the best value for k
k_means = KMeans(n_clusters= 4)

In [128]:
# Fit the K-Means model using the scaled data
k_means.fit(df_scaled_indexed)

KMeans(n_clusters=4)

In [129]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_means_predict = k_means.predict(df_scaled_indexed)

# Print the resulting array of cluster values.
k_means_predict


array([3, 3, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 3, 0, 3, 0, 0, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 1, 3, 0, 0, 2, 0, 0, 0, 0],
      dtype=int32)

In [130]:
# Create a copy of the DataFrame
predicted_scaled_df = df_scaled_indexed.copy()
predicted_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [131]:
# Add a new column to the DataFrame with the predicted clusters
predicted_scaled_df["predicted clusters"] = k_means_predict

# Display sample data
predicted_scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted clusters  
coin_id                                                       
bitcoin                        -0.251637                   3  
ethereum                       -0.199352                   3  
tether                         -0.282061                   0  
ripple                         -0.295546                   0  
bitcoin-cash                   -0.270317                   3

In [132]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
predicted_scaled_df.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "predicted clusters" ,
    title = " pedicted clusters"
)


:NdOverlay   [predicted clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In this code, I am using the PCA model with `n_components=3` to perform dimensionality reduction on `predicted_scaled_df`, and then viewing the first five rows of the transformed data (`pca_fit`). Additionally, you are retrieving the explained variance to understand how much information each principal component retains.

1. `pca = PCA(n_components=3)`: You create a PCA model instance with `n_components=3`, indicating that you want to reduce the dimensionality of the data to three principal components.

2. `pca_fit = pca.fit_transform(predicted_scaled_df)`: The PCA model is fitted to the `predicted_scaled_df` DataFrame using the `fit_transform()` method. The original high-dimensional data is transformed into a lower-dimensional space containing three principal components. The result is stored in the `pca_fit` variable.

3. `pca_fit[:5]`: You view the first five rows of the transformed data (`pca_fit`) to observe the reduced feature data after PCA. This step allows you to inspect the transformed data and understand how the principal components represent the information from the original features.

4. Retrieve Explained Variance:
   - The explained variance tells you how much information (variance) can be attributed to each principal component.
   - It is obtained from the `explained_variance_ratio_` attribute of the PCA model.

```python
# Retrieve explained variance
explained_variance = pca.explained_variance_ratio_
```

The `explained_variance` will be an array containing the explained variance ratio for each of the three principal components. The explained variance ratio represents the proportion of the total variance in the data that each principal component explains.

To understand PCA further and explore the concept of explained variance in detail, you can refer to the following resources:

1. Scikit-learn documentation on PCA:
   - https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html

2. Towards Data Science article on PCA:
   - https://towardsdatascience.com/a-one-stop-shop-for-principal-component-analysis-5582fb7e0a9c

3. PCA Explained (Visual Guide) by StatQuest with Josh Starmer:
   - Video: https://www.youtube.com/watch?v=_UVHneBUBW0
   - Written explanation: https://statquest.org/pca-principal-component-analysis-step-by-step/

In [133]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components= 3)

In [134]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_fit = pca.fit_transform(predicted_scaled_df)
# View the first five rows of the DataFrame. 
pca_fit[:5] 

array([[ 2.05913887, -0.53668446, -0.43823774],
       [ 1.95640089, -0.41082149, -1.0169655 ],
       [-0.99014153, -0.45422783,  0.69002131],
       [-0.9731886 , -0.49464992,  0.55123402],
       [ 3.33036692, -1.02231141, -0.50134143]])

In [135]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
# Retrieve explained variance
explained_variance = pca.explained_variance_ratio_
total_explained_var =  explained_variance[0] + explained_variance[1] + explained_variance[2]
print(f" {round(total_explained_var * 100 , 2)}% is the total var")

 88.92% is the total var


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** the total of the 3 PCA is 88.92%

In [136]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(pca_fit , columns=["PC1", "PC2", "PC3"])
pca_df.head()

pca_df["coin_id"] = df_market_data.index

pca_df.set_index("coin_id")
# Creating a DataFrame with the PCA data


# Copy the crypto names from the original data


# Set the coinid column as index


# Display sample data


PC1       PC2       PC3
coin_id                                           
bitcoin               2.059139 -0.536684 -0.438238
ethereum              1.956401 -0.410821 -1.016966
tether               -0.990142 -0.454228  0.690021
ripple               -0.973189 -0.494650  0.551234
bitcoin-cash          3.330367 -1.022311 -0.501341
binancecoin           2.527411 -0.422464 -0.287137
chainlink             2.603960 -0.387534 -2.182795
cardano               2.305615 -0.280835 -1.175874
litecoin              1.829981 -0.605545 -0.800958
bitcoin-cash-sv      -0.857818 -0.778493  0.401555
crypto-com-chain     -2.066073 -0.336958  0.387962
usd-coin             -1.001762 -0.460074  0.699153
eos                  -1.147255 -0.729127  0.418427
monero                3.743094  0.236363  0.282297
tron                 -0.752140 -0.407313  0.331470
tezos                 1.375125 -0.796222 -1.733582
okb                  -1.892854 -0.017249  0.222054
stellar              -1.176305 -0.535083 -0.061559
cosmos                2.360880 -0.241337 -1.343141
cdai                 -0.975439 -0.533392  0.719632
neo                  -1.564505 -0.425837  0.186023
wrapped-bitcoin       2.041760 -0.541215 -0.434218
leo-token            -1.461865 -0.463200  0.616967
huobi-token          -0.927665 -0.429001  0.474541
nem                  -0.715809  0.620327  1.487186
binance-usd          -0.979863 -0.470508  0.705698
iota                 -0.930645 -0.797473 -0.084370
vechain              -2.351505 -0.671871 -0.124814
zcash                 2.013153 -0.375257 -1.062802
theta-token          -1.431482  2.650660  1.539466
dash                 -0.873617 -0.644158 -0.141833
ethereum-classic     -0.935719 -0.606902  0.226281
ethlend              -2.226657  7.893463 -4.304048
maker                 1.582127 -0.360428 -0.861539
havven               -2.134677  0.741981 -1.359790
omisego              -0.787830  0.116679  1.427688
celsius-degree-token  4.336717  5.127531  5.344202
ontology             -2.277983 -0.752627 -0.612321
ftx-token            -0.583903 -0.598662  0.280874
true-usd             -0.965452 -0.477553  0.708552
digibyte             -1.083578 -0.321990  0.826044

In [140]:
pca_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41 entries, 0 to 40
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   PC1      41 non-null     float64
 1   PC2      41 non-null     float64
 2   PC3      41 non-null     float64
 3   coin_id  41 non-null     object 
dtypes: float64(3), object(1)
memory usage: 1.4+ KB


---

### Find the Best Value for k Using the PCA Data

In [146]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))
inertia = []
pca_df_indexed = pca_df.set_index("coin_id")
pca_df_indexed.head()

PC1       PC2       PC3
coin_id                                   
bitcoin       2.059139 -0.536684 -0.438238
ethereum      1.956401 -0.410821 -1.016966
tether       -0.990142 -0.454228  0.690021
ripple       -0.973189 -0.494650  0.551234
bitcoin-cash  3.330367 -1.022311 -0.501341

In [147]:
# Create a for loop to compute the inertia with each possible value of k
for i in k :
    k_model = KMeans(n_clusters= i , random_state= 3)
    k_model.fit(pca_df_indexed)
    inertia.append(k_model.inertia_)
inertia[:5]


[325.4084233851049,
 195.309950557572,
 108.66475812556446,
 39.88332402229196,
 28.98624018250908]

In [148]:
# Create a dictionary with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame({
    "k" : k,
    "inertia" : inertia
})
df_elbow_pca



k     inertia
0   1  325.408423
1   2  195.309951
2   3  108.664758
3   4   39.883324
4   5   28.986240
5   6   21.263140
6   7   16.169572
7   8   13.426848
8   9    9.976960
9  10    8.135239

In [150]:
# Plot a line chart with all the inertia values computed with 
elbow_pca_plot = df_elbow_pca.hvplot.line(
    x = "k",
    y = "inertia",
    title = "PCA Elbow Curve"
)
elbow_pca_plot
# the different values of k to visually identify the optimal value for k.


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Four is the best value for k


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no it doesnt

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [151]:
# Initialize the K-Means model using the best value for k
k_means = KMeans( n_clusters=4, random_state=3)

In [152]:
# Fit the K-Means model using the PCA data
k_means.fit(pca_df_indexed)

KMeans(n_clusters=4, random_state=3)

In [ ]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k
# Print the resulting array of cluster values.


In [ ]:
# Create a copy of the DataFrame with the PCA data


# Add a new column to the DataFrame with the predicted clusters


# Display sample data


In [ ]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [ ]:
# Composite plot to contrast the Elbow curves

In [ ]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 